File: sample-videochat-webrtc/sample-videochat-webrtc/ViewControllers/Settings/VideoSettingsViewController/VideoSettingsViewController.m
//
//  SettingsViewController.m
//  sample-videochat-webrtc
//
//  Created by Andrey Ivanov on 21.06.15.
//  Copyright (c) 2015 QuickBlox Team. All rights reserved.
//

#import "VideoSettingsViewController.h"
#import "Settings.h"

#import "SettingCell.h"
#import "SettingSliderCell.h"
#import "SettingSwitchCell.h"

#import "SwitchItemModel.h"
#import "SliderItemModel.h"
#import "SettingsSectionModel.h"

typedef NS_ENUM(NSUInteger, VideoSettingsSectionType) {
    
    VideoSettingsSectionCameraPostion = 0,
    VideoSettingsSectionVideoCodec = 1,
    VideoSettingsSectionSupportedFormats = 2,
    VideoSettingsSectionVideoFrameRate = 3,
    VideoSettingsSectionBandwidth = 4,
    VideoSettingsSectionController = 5
};

@implementation VideoSettingsViewController

- (NSString *)titleForSection:(NSUInteger)section {
    
    switch (section) {
        case VideoSettingsSectionCameraPostion:
            return @"Switch camera position";
        case VideoSettingsSectionVideoCodec:
            return @"Video codec";
        case VideoSettingsSectionSupportedFormats:
            return @"Video formats";
        case VideoSettingsSectionVideoFrameRate:
            return @"Frame rate";
        case VideoSettingsSectionBandwidth:
            return @"Bandwidth";
        case VideoSettingsSectionController:
            return @"Call controller";
    }
    
    return nil;
}

- (NSArray *)videoFormatModelsWithCameraPositon:(AVCaptureDevicePosition)cameraPosition {
    //Grab supported formats
    
    NSArray *formats = [QBRTCCameraCapture formatsWithPosition:cameraPosition];
    
    NSMutableArray *videoFormatModels = [NSMutableArray arrayWithCapacity:formats.count];
    for (QBRTCVideoFormat *videoFormat in formats) {
        
        BaseItemModel *videoFormatModel = [[BaseItemModel alloc] init];
        videoFormatModel.title = [NSString stringWithFormat:@"%tux%tu", videoFormat.width, videoFormat.height];
        videoFormatModel.data = videoFormat;
        [videoFormatModels addObject:videoFormatModel];
    }
    
    return videoFormatModels;
}

- (void)configure {
    
    //Camera position section
    __weak __typeof(self)weakSelf = self;
    [self addSectionWith:VideoSettingsSectionCameraPostion items:^NSArray *(NSString *sectionTitle) {
        
        //Camera position section
        SwitchItemModel *switchItem = [[SwitchItemModel alloc] init];
        switchItem.title = @"Back Camera";
        
        switchItem.on = (weakSelf.settings.preferredCameraPostion == AVCaptureDevicePositionBack);
        
        return @[switchItem];
    }];
    //Video codecs
    [self addSectionWith:VideoSettingsSectionVideoCodec items:^NSArray *(NSString *sectionTitle) {
        
        BaseItemModel *vp8Model = [[BaseItemModel alloc] init];
        vp8Model.title = @"VP8";
        vp8Model.data = @(QBRTCVideoCodecVP8);
        
        BaseItemModel *h264BaselineModel = [[BaseItemModel alloc] init];
        h264BaselineModel.title = @"H264Baseline";
        h264BaselineModel.data = @(QBRTCVideoCodecH264Baseline);
        
        BaseItemModel *h264HighModel = [[BaseItemModel alloc] init];
        h264HighModel.title = @"H264High";
        h264HighModel.data = @(QBRTCVideoCodecH264High);
        
        [weakSelf selectSection:VideoSettingsSectionVideoCodec index:(NSUInteger)weakSelf.settings.mediaConfiguration.videoCodec];
        
        return @[vp8Model, h264BaselineModel, h264HighModel];
    }];
    //Supported video formats section
    [self addSectionWith:VideoSettingsSectionSupportedFormats items:^NSArray *(NSString *sectionTitle) {
        
        AVCaptureDevicePosition position = weakSelf.settings.preferredCameraPostion;
        NSArray *videoFormats = [weakSelf videoFormatModelsWithCameraPositon:position];
        
        NSArray *formats = [QBRTCCameraCapture formatsWithPosition:position];
        //Select index path
        NSUInteger idx = [formats indexOfObject:weakSelf.settings.videoFormat];
        [weakSelf selectSection:VideoSettingsSectionSupportedFormats index:idx];
        
        return videoFormats;
    }];
    //Frame rate
    [self addSectionWith:VideoSettingsSectionVideoFrameRate items:^NSArray *(NSString *sectionTitle) {
        
        SliderItemModel *frameRateSlider = [[SliderItemModel alloc] init];
        frameRateSlider.title = @"30";
        frameRateSlider.minValue = 2;
        frameRateSlider.currentValue = weakSelf.settings.videoFormat.frameRate;
        frameRateSlider.maxValue = 30;
        
        return @[frameRateSlider];
    }];
    //Video bandwidth
    [self addSectionWith:VideoSettingsSectionBandwidth items:^NSArray *(NSString *sectionTitle) {
        
        SliderItemModel *bandwidthSlider = [[SliderItemModel alloc] init];
        bandwidthSlider.title = @"30";
        bandwidthSlider.minValue = 0;
        bandwidthSlider.currentValue = weakSelf.settings.mediaConfiguration.videoBandwidth;
        bandwidthSlider.maxValue = 2000;
        
        return @[bandwidthSlider];
    }];
    //Call controller
    [self addSectionWith:VideoSettingsSectionController items:^NSArray *(NSString *sectionTitle) {
        
        SwitchItemModel *switchItem = [[SwitchItemModel alloc] init];
        switchItem.title = @"Prefer Metal for drawing";
        
        switchItem.on = QBRTCRemoteVideoView.preferMetal;
        
        return @[switchItem];
    }];
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    switch (indexPath.section) {
            
        case VideoSettingsSectionVideoCodec:
            [self updateSelectionAtIndexPath:indexPath];
            // need to request available formats 1 more time for new codec
            // but we need to save it first
            [self applySettings];
            [self.settings applyConfig];
            [self reloadVideoFormatSectionForPosition:self.settings.preferredCameraPostion];
            break;
            
        case VideoSettingsSectionSupportedFormats: {
            
            [self updateSelectionAtIndexPath:indexPath];
            break;
        }
    }
}

#pragma mark - SettingsCellDelegate

- (void)cell:(BaseSettingsCell *)cell didChageModel:(BaseItemModel *)model {
    
    if ([model isKindOfClass:[SwitchItemModel class]]) {
        NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
        BOOL isOn = ((SwitchItemModel *)model).on;
        switch (indexPath.section) {
            case VideoSettingsSectionVideoCodec:
                [self reloadVideoFormatSectionForPosition:isOn ? AVCaptureDevicePositionBack : AVCaptureDevicePositionFront];
                break;
            case VideoSettingsSectionController: {
                QBRTCRemoteVideoView.preferMetal = isOn;
                break;
            }
        }
    }
}

#pragma mark - Helpers

- (void)reloadVideoFormatSectionForPosition:(AVCaptureDevicePosition)position {
    
    NSArray *videoFormatModels = [self videoFormatModelsWithCameraPositon:position];
    
    SettingsSectionModel *section = [self sectionWith:VideoSettingsSectionSupportedFormats];
    section.items = videoFormatModels;
    NSArray *formats = [QBRTCCameraCapture formatsWithPosition:position];
    
    NSString *title = [self titleForSection:VideoSettingsSectionSupportedFormats];
    NSIndexPath *oldIdnexPath = self.selectedIndexes[title];
    //Select index path
    
    NSUInteger idx = section.items.count - 1;
    if (idx >= oldIdnexPath.row) {
        
        BaseItemModel *videoFormatModel = section.items[oldIdnexPath.row];
        QBRTCVideoFormat *videoFormat = videoFormatModel.data;
        
        idx = [formats indexOfObject:videoFormat];
    }
    
    [self selectSection:VideoSettingsSectionSupportedFormats index:idx];
    
    NSIndexSet *sectionToReload = [NSIndexSet indexSetWithIndex:VideoSettingsSectionSupportedFormats];
    [self.tableView reloadSections:sectionToReload withRowAnimation:UITableViewRowAnimationFade];
}

- (void)applySettings {
    
    //APPLY SETTINGS
    
    //Preferred camera positon
    SwitchItemModel *cameraPostion = (id)[self modelWithIndex:0 section:VideoSettingsSectionCameraPostion];
    self.settings.preferredCameraPostion = cameraPostion.on ? AVCaptureDevicePositionBack : AVCaptureDevicePositionFront;
    
    //Video codec
    NSIndexPath *videoCodecIndexPath = [self indexPathAtSection:VideoSettingsSectionVideoCodec];
    BaseItemModel *videoCodec = [self modelWithIndex:videoCodecIndexPath.row section:videoCodecIndexPath.section];
    self.settings.mediaConfiguration.videoCodec = (QBRTCVideoCodec)[(NSNumber *)videoCodec.data integerValue];
    
    //Supported format
    NSIndexPath *supportedFormatIndexPath = [self indexPathAtSection:VideoSettingsSectionSupportedFormats];
    BaseItemModel *format = [self modelWithIndex:supportedFormatIndexPath.row section:supportedFormatIndexPath.section];
    QBRTCVideoFormat *videoFormat = format.data;
    
    //Frame rate
    SettingsSectionModel *frameRate = [self sectionWith:VideoSettingsSectionVideoFrameRate];
    SliderItemModel *frameRateSlider = frameRate.items.firstObject;
    
    //bandwidth
    SettingsSectionModel *bandwidth = [self sectionWith:VideoSettingsSectionBandwidth];
    SliderItemModel *bandwidthSlider = bandwidth.items.firstObject;
    
    self.settings.mediaConfiguration.videoBandwidth = bandwidthSlider.currentValue;
    
    self.settings.videoFormat =
    [QBRTCVideoFormat videoFormatWithWidth:videoFormat.width
                                    height:videoFormat.height
                                 frameRate:frameRateSlider.currentValue
                               pixelFormat:QBRTCPixelFormat420f];
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Settings/VideoSettingsViewController/VideoSettingsViewController.swift
//
//  VideoSettingsViewController.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 12/11/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import AVKit
import QuickbloxWebRTC

enum VideoSettingsSectionType: Int {
    case cameraPostion
    case videoCodec
    case supportedFormats
    case videoFrameRate
    case bandwidth
    case sectionController
}

class VideoSettingsViewController: BaseSettingsViewController {
    
    //MARK: - Overrides
    override func title(forSection section: Int) -> String {
        switch section {
        case VideoSettingsSectionType.cameraPostion.rawValue:
            return "Switch camera position"
        case VideoSettingsSectionType.videoCodec.rawValue:
            return "Video codec"
        case VideoSettingsSectionType.supportedFormats.rawValue:
            return "Video formats"
        case VideoSettingsSectionType.videoFrameRate.rawValue:
            return "Frame rate"
        case VideoSettingsSectionType.bandwidth.rawValue:
            return "Bandwidth"
        case VideoSettingsSectionType.sectionController.rawValue:
            return "Call controller"
        default:
            return ""
        }
    }
    
    //MARK: - Overrides
    override func configure() {
        let settings = Settings()
        //Camera position section
        addSection(with: VideoSettingsSectionType.cameraPostion.rawValue, items: { sectionTitle in
            //Camera position section
            let switchItem = SwitchItemModel()
            switchItem.title = "Back Camera"
            switchItem.on = settings.preferredCameraPostion == .back
            return [switchItem]
        })
        //Video codecs
        addSection(with: VideoSettingsSectionType.videoCodec.rawValue, items: { [weak self] sectionTitle in
            let vp8Model = BaseItemModel()
            vp8Model.title = "VP8"
            vp8Model.data = QBRTCVideoCodec.VP8
            
            let h264BaselineModel = BaseItemModel()
            h264BaselineModel.title = "H264Baseline"
            h264BaselineModel.data = QBRTCVideoCodec.h264Baseline
            
            let h264HighModel = BaseItemModel()
            h264HighModel.title = "H264High"
            h264HighModel.data = QBRTCVideoCodec.h264High
            
            let videoCodec = settings.mediaConfiguration.videoCodec
            self?.selectSection(VideoSettingsSectionType.videoCodec.rawValue, index: Int(videoCodec.rawValue))
            
            
            return [vp8Model, h264BaselineModel, h264HighModel]
        })
        //Supported video formats section
        addSection(with: VideoSettingsSectionType.supportedFormats.rawValue, items: { [weak self] sectionTitle in
            var videoFormats = [BaseItemModel]()
            let position = settings.preferredCameraPostion
            guard let videoFormatsModels = self?.videoFormatModels(withCameraPositon: position)  else {
                return videoFormats
            }
            videoFormats = videoFormatsModels
            var formats = QBRTCCameraCapture.formats(with: position)
            formats = formats.filter({ $0.width <= 1024 })
            //Select index path
            let idx: Int = (formats as NSArray).index(of: settings.videoFormat as Any)
            self?.selectSection(VideoSettingsSectionType.supportedFormats.rawValue, index: idx)
            
            return videoFormats
        })
        //Frame rate
        addSection(with: VideoSettingsSectionType.videoFrameRate.rawValue, items: { sectionTitle in
            let frameRateSlider = SliderItemModel()
            frameRateSlider.title = "30"
            frameRateSlider.minValue = 2
            frameRateSlider.maxValue = 30
            let currentValue = settings.videoFormat.frameRate
            frameRateSlider.currentValue = currentValue
            
            return [frameRateSlider]
        })
        //Video bandwidth
        addSection(with: VideoSettingsSectionType.bandwidth.rawValue, items: { sectionTitle in
            var currentValue = 30
            let bandwidthSlider = SliderItemModel()
            bandwidthSlider.title = "30"
            bandwidthSlider.minValue = 0
            let currValue = settings.mediaConfiguration.videoBandwidth
            currentValue = currValue
            
            bandwidthSlider.currentValue = UInt(bitPattern: currentValue)
            bandwidthSlider.maxValue = 2000
            return [bandwidthSlider]
        })
        addSection(with: VideoSettingsSectionType.sectionController.rawValue, items: { sectionTitle in
            let switchItem = SwitchItemModel()
            switchItem.title = "Prefer Metal for drawing"
            switchItem.on = QBRTCRemoteVideoView.preferMetal
            
            return [switchItem]
        })
    }
    
    // MARK: - UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        switch indexPath.section {
        case Int(VideoSettingsSectionType.videoCodec.rawValue):
            updateSelection(at: indexPath)
            // need to request available formats 1 more time for new codec
            // but we need to save it first
            applySettings()
            
        case Int(VideoSettingsSectionType.supportedFormats.rawValue):
            updateSelection(at: indexPath)
        default:
            break
        }
    }
    
    // MARK: - SettingsCellDelegate
    override func cell(_ cell: BaseSettingsCell, didChageModel model: BaseItemModel) {
        guard let model = model as? SwitchItemModel else {
            return }
        guard let indexPath = tableView.indexPath(for: cell) else {
            return
        }
        switch indexPath.section {
        case Int(VideoSettingsSectionType.videoCodec.rawValue):
            reloadVideoFormatSection(for: model.on ? .back : .front)
            
        case Int(VideoSettingsSectionType.sectionController.rawValue):
            QBRTCRemoteVideoView.preferMetal = model.on
        default:
            break
        }
    }
    
    override func applySettings() {
        //APPLY SETTINGS
        //Preferred camera positon
        let settings = Settings()
        guard let cameraPostion = model(with: 0, section: VideoSettingsSectionType.cameraPostion.rawValue) as? SwitchItemModel else {
            return }
        settings.preferredCameraPostion = cameraPostion.on ? .back : .front
        
        //Video codec
        guard let videoCodecIndexPath = indexPath(atSection: VideoSettingsSectionType.videoCodec.rawValue),
            let videoCodec = model(with: videoCodecIndexPath.row, section: videoCodecIndexPath.section),
            let videoCodecData = videoCodec.data as? QBRTCVideoCodec  else {
                return
        }
        settings.mediaConfiguration.videoCodec = videoCodecData
        
        //Supported format
        guard let supportedFormatIndexPath = indexPath(atSection: VideoSettingsSectionType.supportedFormats.rawValue),
            let format = model(with: supportedFormatIndexPath.row, section: supportedFormatIndexPath.section),
            let videoFormat = format.data as? QBRTCVideoFormat else {
                return
        }
        
        //Frame rate
        guard let frameRate = section(with: VideoSettingsSectionType.videoFrameRate.rawValue),
            let frameRateSlider = frameRate.items.first as? SliderItemModel else {
                return
        }
        
        //bandwidth
        guard let bandwidth = section(with: VideoSettingsSectionType.bandwidth.rawValue),
            let bandwidthSlider = bandwidth.items.first as? SliderItemModel else {
                return
        }
        settings.mediaConfiguration.videoBandwidth = Int(bandwidthSlider.currentValue)
        settings.videoFormat = QBRTCVideoFormat.init(width: videoFormat.width,
                                                     height: videoFormat.height,
                                                     frameRate: frameRateSlider.currentValue,
                                                     pixelFormat: QBRTCPixelFormat.format420f)
        
        settings.applyConfig()
        settings.saveToDisk()
        reloadVideoFormatSection(for: settings.preferredCameraPostion)
    }
    
    //MARK: - Internal Methods
    func videoFormatModels(withCameraPositon cameraPosition: AVCaptureDevice.Position) -> [BaseItemModel]? {
        //Grab supported formats
        var formats = QBRTCCameraCapture.formats(with: cameraPosition)
        formats = formats.filter({ $0.width <= 1024 })
        var videoFormatModels = [BaseItemModel]()
        for videoFormat in formats {
            let videoFormatModel = BaseItemModel()
            videoFormatModel.title = "\(videoFormat.width)x\(videoFormat.height)"
            videoFormatModel.data = videoFormat
            videoFormatModels.append(videoFormatModel)
        }
        return videoFormatModels
    }
    
    // MARK: - Helpers
    func reloadVideoFormatSection(for position: AVCaptureDevice.Position) {
        guard let videoFormatModels = self.videoFormatModels(withCameraPositon: position),
            let section = self.section(with: VideoSettingsSectionType.supportedFormats.rawValue) else {
                return }
        section.items = videoFormatModels
        let formats = QBRTCCameraCapture.formats(with: position)
        let title = self.title(forSection: VideoSettingsSectionType.supportedFormats.rawValue)
        
        var idx: Int = section.items.count
        if let oldIdnexPath = selectedIndexes[title] {
            //Select index path
            if idx >= oldIdnexPath.row {
                let videoFormatModel = section.items[oldIdnexPath.row]
                let videoFormat = videoFormatModel.data as? QBRTCVideoFormat
                if let videoFormat = videoFormat {
                    idx = (formats as NSArray).index(of: videoFormat)
                }
            }
        }
        selectSection(VideoSettingsSectionType.supportedFormats.rawValue, index: idx)
        let sectionToReload = NSIndexSet(index: Int(VideoSettingsSectionType.supportedFormats.rawValue))
        tableView.reloadSections(sectionToReload as IndexSet, with: .fade)
    }
}
